pkg/cafs/reader.go

```diff
@@ -200,73 +207,93 @@ func calculateKeyAndOffset(off int64, leafSize uint32) (index int64, offset int6
 	return
 }
 
-func (r *chunkReader) ReadAt(p []byte, off int64) (n int, err error) {
+func (r *chunkReader) ReadAt(p []byte, off int64) (totread int, err error) {
+
+	readLeaf := func(k Key) (*leafBuffer, error) {
+		rdr, e := r.fs.Get(context.Background(), k.StringWithPrefix(r.prefix))
+		if e != nil {
+			return nil, e
+		}
+		lb := r.leafPool.get()
+		var n int
+		for {
+			m, e := rdr.Read(lb.slice[n:cap(lb.slice)])
+			if m < 0 {
+				panic(fmt.Errorf("read negative number of bytes"))
+			}
+			n += m
+			lb.slice = lb.slice[:n]
```

Comment:
I have some questions around this, not sure why you need to do ```lb.slice = lb .slice[:n]```

Reply:
plan to resolve this:   use `len(lb.slice)` instead of `n` .. this'll add some compute time while making the code more readable.

incidentally, this implementation was originally lifted from [the implementation](https://golang.org/src/bytes/buffer.go) of [`bytes.Buffer.ReadFrom`](https://golang.org/pkg/bytes/#Buffer.ReadFrom), which is what the original `ioutil.ReadAll` call it replaces uses, and the copyover was done without too much thought.